boot/freeldr/freeldr/arch/realmode/helpers_all.inc

```diff
@@ -0,0 +1,154 @@
+/*
+ * PROJECT:     FreeLoader
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     Real mode helper code for all architectures
+ * COPYRIGHT:   Copyright 2024-2025 Daniel Victor <ilauncherdeveloper@gmail.com>
+ */
+
+#define RELOCATOR_BLOCK_SIZE            HEX(200)
+#define RELOCATOR_BLOCK_SIZE_SEGMENT    (RELOCATOR_BLOCK_SIZE / 16)
+
+#define RELOCATOR_FUNCTION_BASE         FREELDR_BASE
+
+#define RELOCATOR_PROGRAM_SIZE          (MEMORY_MARGIN - FREELDR_BASE)
+#define RELOCATOR_PROGRAM_SIZE_BLOCKS   ((RELOCATOR_PROGRAM_SIZE + (RELOCATOR_BLOCK_SIZE - 1)) / RELOCATOR_BLOCK_SIZE)
+#define RELOCATOR_PROGRAM_SIZE_ALIGNED  (RELOCATOR_PROGRAM_SIZE_BLOCKS * RELOCATOR_BLOCK_SIZE)
+
+/* This function will Update BSS_CurrentSegmentBaseAddress and relocate if necessary */
+RelocateFreeLdr:
+    /* Clear direction flag */
+    cld
+
+    /* Call the sub function to get the base address */
+    call .SubCall
+.SubCall:
+    /* Pop the return address into bx. Since the return address is consumed,
+     * this function does not return to UpdateCurrentCodeSegment */
+    pop bx
+
+    /* Calculate IP base address */
+    sub bx, offset .SubCall - RELOCATOR_FUNCTION_BASE
+
+    /* Convert IP base address to segment value */
+    shr bx, 4
+
+    /* Load CS base address into ax */
+    mov ax, cs
+
+    /* Add CS base address with IP base address */
+    add ax, bx
+
+    /* Store the calculated base address (low 16 bits) */
+    mov word ptr fs:[BSS_CurrentSegmentBaseAddress], ax
+
+.RelocateFrLdr16:
+    /* If the current base address is the same as FREELDR_BASE then it's not necessary to relocate */
+    cmp ax, FREELDR_BASE / 16
+    jz .NotNecessaryRelocation
+
+    /* Clear counter register */
+    xor cx, cx
+
+    /* Setup source segment */
+    mov ds, ax
+
+    /* Move the realmode part of FRLDR16 code to TEMPCODE16_BASE */
+    mov di, TEMPCODE16_BASE
+    mov si, offset RELOCATOR_FUNCTION_START - RELOCATOR_FUNCTION_BASE
+    mov cl, offset RELOCATOR_FUNCTION_END - RELOCATOR_FUNCTION_START
+    rep movsb
+
+    ljmp16 TEMPCODE16_BASE / 16, 0
```

Comment:
Here you could add in comment what the effect of this jump is. (As I understand the code it should jump to the relocated `RELOCATOR_FUNCTION_START`, right?)